Implement module collection mode setting for hooks

rokm opened the pull request:
Implement a mechanism for controlling the collection mode of modules and packages, with granularity ranging from top-level packages to individual sub-modules. Therefore, the hooks can now specify whether the hooked package should be collected as
byte-compiled .pyc modules into embedded PYZ archive (the default behavior), or as source .py files collected as external data files (without corresponding modules in the PYZ archive).

The latter option should let us avoid unnecessary .pyc module collection when the source files are required by the code, or
work around the situations where having a .pyc module in PYZ archive causes issues due to FrozenImporter's incompatibility with sys.path manipulation that some packages attempt to perform.

This feature adds a new optional global hook variable, called `module_collection_mode`. The value can be either a string ("py" or "pyc") or a dictionary of module names and setting strings.

In the case of a string, the setting affects the hooked module or a package, and is applied recursively to all sub-packages and sub-modules, unless another hook overrides it.

The dictionary setting allows a hook to specify different settings for the package and it subpackages, or even different settings for other packages.

A corresponding `set_module_collection_mode` method has been added to the `hook_api` object for adjusting the collection mode from within the `hook()` function.

The `Analysis` object can now also be passed a dictionary via an optional `module_collection_mode` argument; the corresponding settings are applied last, which allows advanced users to both supplement and override the settings made by the hooks.

rokm committed:
building & hooks: implement module collection mode setting

Implement a mechanism for controlling the collection mode of
modules and packages, with granularity ranging from top-level
packages to individual sub-modules. Therefore, the hooks can
now specify whether the hooked package should be collected as
byte-compiled .pyc modules into embedded PYZ archive (the
default behavior), or as source .py files collected as external
data files (without corresponding modules in the PYZ archive).

The latter option should let us avoid unnecessary .pyc module
collection when the source files are required by the code, or
work around the situations where having a .pyc module in
PYZ archive causes issues due to FrozenImporter's incompatibility
with sys.path manipulation that some packages attempt to perform.

This feature adds a new optional global hook variable, called
`module_collection_mode`. The value can be either a string
("py" or "pyc") or a dictionary of module names and setting
strings.

In the case of a string, the setting affects the hooked module
or a package, and is applied recursively to all sub-packages and
sub-modules, unless another hook overrides it.

The dictionary setting allows a hook to specify different settings
for the package and it subpackages, or even different settings
for other packages.

A corresponding `set_module_collection_mode` method has been
added to the `hook_api` object for adjusting the collection
mode from within the `hook()` function.

The `Analysis` object can now also be passed a dictionary via
an optional `module_collection_mode` argument; the corresponding
settings are applied last, which allows advanced users to both
supplement and override the settings made by the hooks.

rokm commented:
This originally started as a `collect_as_source` boolean global hook variable (translated into dict of boolean settings behind the scenes) that affected the hooked package (and its contents). 

The value has then been changed to `'py'` and `'pyc'`, because we might also want to also handle `'py+pyc'` (collecting both pyc modules and source files; which can currently be achieved by `collect_all` or `collect_submodules` + `collect_data_files`). Or some other special cases, such as `pyc-without-pyz`.

Being able to also specify multiple settings from a single hook via dictionary allows a hook to provide different settings for its contents without having to implement hooks for sub-packages. I'm not sure if we'll ever run into situation where that would be useful, though. On the other hand, being able to specify mode for other packages may be useful in cases such as `torch` where the `torch` hook could set the collect-as-source mode for all packages that use its JIT functionality.

bwoodsend commented:
> The value has then been changed to 'py' and 'pyc', because we might also want to also handle 'py+pyc' (collecting both pyc modules and source files; which can currently be achieved by collect_all or collect_submodules + collect_data_files). Or some other special cases, such as pyc-without-pyz.

I would have thought that, even though `.py` files will be automatically compiled to `.pyc`, it would still be worthwhile to collect the `.pyc` files too at least in `--onefile` so that the compilation doesn't run every time and hinder startup time. If we choose to go against that, it might be worth setting `PYTHONDONTWRITEBYTECODE` (or the `-B` flag).

rokm commented:
Fair enough. My main motivation here is
a) to be able to bypass PYZ on per module/package basis, i.e., without resorting to `noarchive=True` for the whole build
b) to be able to collect source .py files based on modulegraph analysis, i.e., without having to resort to `collect_all` or `collect_data_files(..., include_py_files=True)`.

So perhaps we should consider the following scenarios / mode strings:
* `pyz`: (the default): byte-compiled .pyc is collected into PYZ archive (unless `noarchive=True` is used)
* `pyc`: byte-compiled .pyc is collected as external data file (e.g, to bypass the PYZ/FrozenImporter issues)
* `py`: source .py is collected as external data file (and no .pyc is collected)
* `pyz+py`, `py+pyz`: collect source .py as external data file and collect byte-compiled .pyc into PYZ
* `pyc+py`, ˙py+pyc˙: collect source .py as external data file and byte-compiled .pyc as external data file

For `pyc+py`, is it enough to collect files side-by-side, or do we need to place the .pyc into `__pycache__`?

I think we should probably be setting `PYTHONDONTWRITEBYTECODE` by default anyway, because frozen bundles are supposed to be immutable. 

I rather dislike the idea of collecting both .py and .pyc, because this creates an opportunity for disconnect between them - i.e., if the .py file is changed for some reason, the .pyc (especially in the PYZ) will not reflect that.

bwoodsend commented:
> For pyc+py, is it enough to collect files side-by-side, or do we need to place the .pyc into __pycache__?

Just monkeying around, it looks like the precompiled file is ignored and recompiled if I try to encourage it to use side-by-side layout instead of the `__pycache__` directory. So I think that we do need the `__pycache__` directory.

```fish
/e/n/p/6945> touch foo.py
/e/n/p/6945> python -m py_compile foo.py 
/e/n/p/6945> find
.
./foo.py
./__pycache__
./__pycache__/foo.cpython-310.pyc

/e/n/p/6945> mv __pycache__/foo.cpython-310.pyc .
/e/n/p/6945> rm -r __pycache__/

/e/n/p/6945> python -c 'import foo'
/e/n/p/6945> find
.
./foo.cpython-310.pyc
./foo.py
./__pycache__
./__pycache__/foo.cpython-310.pyc
```

> I rather dislike the idea of collecting both .py and .pyc, because this creates an opportunity for disconnect between them - i.e., if the .py file is changed for some reason, the .pyc (especially in the PYZ) will not reflect that.

Shall we go for not supporting `.py` + `.pyc` until someone demonstrates a reasonable use case that has enough raw `.py` files that the bytecode compile time really does noticeably hinder startup time? Your mode strings model leaves us room to add modes/mode combinations without breaking things.

rokm commented:
Sounds good to me. Now that I think about it, we fall into exactly this situation (side-by-side collection, and pyc being ignored) when sources are collected via `collect_all` or `collect_data_files` and `noarchive=True` is used, but that should be a corner case rather than a norm...

rokm committed:
building: more module collection modes, consolidate noarchive codepath

Map the module collection mode strings into (combinations of)
integer flags that control try basic collection modes:
- collect a pyc into PYZ archive
- collect a pyc as a data file
- collect a py as a data file

Consolidate the `noarchive=True` codepath into module collection
mode, where "collect a pyc into PYZ archive" flag is swapped
for a "collect a pyc as a data file".

The new collection mode also implicitly fixes couple of minor
annoyances of the `noarchive=True` mode:
- the user-writable paths containing python source code are not
  littered with pyc/pyo files anymore; all pycs are now gathered
  in build directory
- the name of pycs in local build directory are not mangled anymore
  (was previously the case for pycs that could not be written to
  their original locations due to lack of permissions)
- the pycs have code paths stripped from them, same as in
  noarchive=False mode

rokm committed:
utils: remove compile_py_files helper

The only remaining use is in `PYZ.__init__`, and that can be
replaced with a loop that uses the new `compile_pymodule` helper.

This change, however, requires `get_boostrap_modules()` helper
from `PyInstaller.depend˙ to return paths to source `.py`
files instead of non-existing `.pyc` files (the old
`compile_py_files` helper went to great lengths to convert
these back to source file names...).

rokm committed:
building: add support for hash computation in compile_pymodule

If we are compiling py file to pyc, try reading the hash from
the resulting pyc file. If we are given a pyc file, read the hash
from it. If we could not obtain the hash (either the pyc was
timestamp based, or we got code from the code hash), try computing
the hash from the source file.

rokm committed:
CArchive: omit PYC header in PYMODULE ('m', 'M') entries

Store PYMODULE ('m', 'M') entries within the CArchive as pure
marshalled code objects instead of full PYC objects (i.e., strip
the PYC header). These entries are used for bootstrap modules and
runtime hooks, and are loaded and executed by the bootloader. There,
we loaded the blob and skipped the PYC header - so there is no
reason to include the PYC header in the first place...

rokm committed:
building: remove unused fake_pyc_timestamp helper

rokm committed:
building: clear the source hashes in PYC headers of collected .pyc files

When we collect PYMODULE entries into PYZ, we do not preserve PYC
headers and the source hash information within them. Match this
behavior by zero-ing the source hash in PYC headers when modules
are collected as .pyc files, and always disabling the validation
via the header's flags field.

This puts us at odds with PEP-552 that states "Even when validation
is disabled, the hash field should be set correctly, so out-of-band
consistency checkers can verify the up-to-dateness of the pyc", but
I fail to see any benefit in the case of frozen application.

On the other hand, this saves us some CPU cycles in both
`compile_pymodule` and `create_py3_base_library` helpers, because
we can skip the hash computation when we generate the pyc files from
cached code objects.

This commit effectively reverts an earlier commit that added
support for computing source hashes in .pyc compilation helper.

rokm commented:
Alright, here's the second iteration.

We now support `pyz`, `pyc`, `py` and `pyz+py`/`py+pyz` modes, which are internally translated into three integer collection flags ("collect pyc into pyz", "collect pyc as data file", "collect py as data file"). The `noarchive=True` mode is now also consolidated into this codepath, which fixes up some minor problems (like writing .pyc/.pyo files next to source files if location is writable by user, and paths not being stripped from code objects, preventing reproducible builds in `noarchive=True` mode) and allows us to refactor/remove some of the helper code.

bwoodsend reviewed (approved):
I do like f1c7cd65bb3332a83c0237d3f1556b16cbd55715. Why on earth didn't we think of that one before...
